fedbiomed/node/round.py

```diff
@@ -442,7 +463,12 @@ def run_model_training(
                 except Exception as exc:
                     error_message = f"Cannot train model in round: {repr(exc)}"
                     return self._send_round_reply(success=False, message=error_message)
+                try:
```

Comment:
What is the reason for moving these lines here ?

Though it does not break anything, I don't really see a reason.

Moreover, IMO, it makes more sense to keep it where it is currently, aka just before *using* the node state (`results['state_id'] = self._node_state_manager.state_id`). This avoids saving states that won't be used, in case of error between the two.

But that's also OK for me to commit your change, if your think it makes sense to keep it.

Reply:
Discussed together the case where:
- training is True
- test ratio is 1

In this case, this change introduces a bug, as a `state_id` (not None) is created and returned to the researcher, but no state is saved for this `state_id`.

So the options are:
1. restore initial position of block
2. keep position of block but get it out of the `if self.training_plan.training_data_loader is not None:` test
3. add tests for the case where training is True and test_ratio is 1, so it doesn't read state.

IMO:
1. is the most simple and clear
2. is also OK
3. should really really be avoided, as it creates complexity for next-step developers both on node and researcher side: instead of a simple condition "when training is True, then a state is saved by the node", they have to handle the case where test ratio is 1. This is error prone.

Reply:
1. has been done. changes undone.